File: src/segmentation/superpixels.cpp
/*
 * (c)2019-2021, Cris Luengo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "diplib.h"
#include "diplib/segmentation.h"
#include "diplib/generation.h"
#include "diplib/nonlinear.h"
#include "diplib/linear.h"
#include "diplib/math.h"
#include "diplib/morphology.h"

namespace dip {

void Superpixels(
      Image const& in,
      Image& out,
      Random& random,
      dfloat density,
      dfloat compactness,
      String const& method,
      StringSet const& flags
) {
   // Parse input
   DIP_THROW_IF( !in.IsForged(), E::IMAGE_NOT_FORGED );
   DIP_THROW_IF( !in.DataType().IsReal(), E::DATA_TYPE_NOT_SUPPORTED );
   dip::uint nDims = in.Dimensionality();
   DIP_THROW_IF( nDims < 1, E::DIMENSIONALITY_NOT_SUPPORTED );
   bool rectangular = true;
   bool noGaps = false;
   for( auto const& f : flags ) {
      if( f == S::RECTANGULAR ) {
         rectangular = true;
      } else if( f == S::HEXAGONAL ) {
         rectangular = false;
      } else if( f == S::NOGAPS ) {
         noGaps = true;
      } else {
         DIP_THROW_INVALID_FLAG( f );
      }
   }
   if(( nDims > 3 ) || ( nDims < 2 )) {
      rectangular = true;
   }
   String shape = rectangular ? S::RECTANGULAR : ( nDims == 2 ? S::HEXAGONAL : S::FCC );
   // Compute gradient magnitude
   Image gradmag;
   DIP_STACK_TRACE_THIS( GradientMagnitude( in, gradmag ));
   if( !gradmag.IsScalar() ) {
      DIP_STACK_TRACE_THIS( Norm( gradmag, gradmag ));
   }
   // Place seeds
   Image seeds;
   DIP_STACK_TRACE_THIS( CreateRandomGrid( seeds, in.Sizes(), random, density, shape, S::TRANSLATION ));
   MoveToLocalMinimum( seeds, gradmag, seeds );
   // Do the thing
   if( method == S::CW ) {
      StringSet cwFlags;
      cwFlags.emplace( S::LABELS );
      if( noGaps ) {
         cwFlags.emplace( S::NOGAPS );
      }
      DIP_STACK_TRACE_THIS( CompactWatershed( gradmag, seeds, {}, out, 1, compactness, cwFlags ));
   } else {
      DIP_THROW_INVALID_FLAG( method );
   }
}

} // namespace dip

File: dipimage/superpixels.m
%SUPERPIXELS   Generates superpixels (oversegmentation)
%
% SYNOPSIS:
%  image_out = superpixels(image_in,density,compactness,method,flags)
%
% PARAMETERS:
%  density: determines the number of superpixels (average size is 1/density).
%  compactness: determines compactness of superpixels.
%  method: currently only 'CW' is supported (compact watershed superpixels).
%  flags: a cell array of strings, choose from:
%     - 'rectangular' or 'hexagonal': the shape of the superpixels. By default
%       they are rectangular.
%     - 'no gaps': superpixels cover the full image. By default, a 1-pixel
%       gap is left between superpixels.
%
% DEFAULTS:
%  density = 0.005
%  compactness = 1
%  method = 'CW'
%  flags = {}
%
% DIPlib:
%  This function calls the DIPlib function <a href="https://diplib.org/diplib-docs/segmentation.html#dip-Superpixels-Image-CL-Image-L-Random-L-dfloat--dfloat--String-CL-StringSet-CL">dip::Superpixels</a>.

% (c)2019, Cris Luengo.
%
% Licensed under the Apache License, Version 2.0 (the "License");
% you may not use this file except in compliance with the License.
% You may obtain a copy of the License at
%
%    http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
% See the License for the specific language governing permissions and
% limitations under the License.

function out = superpixels(varargin)
out = dip_segmentation('superpixels',varargin{:});
